### DISCUSSION TITLE
zfs rebalance

### DISCUSSION THREAD

**Discussion by dvaezazizi** [openzfs/zfs#13583] [Ideas]
I'm sure this topic has been brought up before, but is there any plan to add the ability to rebalance data after adding vdevs? The performance hit is large and is a pain to deal with at the enterprise level.

**Comment by rincebrain** [comment]
As much as it'd be nice to have, tThe performance hit of doing this, engineering time aside, would be pretty horrific, no matter how you slice it - you'd end up with giant indirection tables like with vdev evacuation on a nontrivial vdev, and all the weird performance implications that ensue there.

I can't say that nobody will ever invent a solution to that that doesn't have gross performance implications, but I think it would involve significantly redesigning the constraints ZFS operates under.

**Comment by Apachez-** [comment]
Today there exists scripts such as:

https://github.com/markusressel/zfs-inplace-rebalancing

However one must trust a 3rd party to run this (along with examinging the script before running it).

At first I thought that "zfs rewrite" would take care of this but changing recordsize for a dataset doesnt seem to have any affect on whatever "zfs rewrite" is doing.

Verifying with "stat <filename>" after running "zfs rewrite -r -v /" shows "IO Block: 131072" when the expected would be "IO Block: 65536" for the usecase of changing recordsize from default 128k to 64k.

So am I doing something wrong here with "zfs rewrite" or will it simply not actually "rewrite" files as I thought it would do (and by that refresh the file to use the current recordsize set for the dataset just as if a new file would have been created)?

In case "zfs rewrite" will do everything except use currently configured recordsize then perhaps there should be a "zfs rebalance" or some syntax for "zfs rewrite" where it would include to also use the new recordsize if such have been changed since the file first was created?

**Comment by robn** [comment]
`zfs-rewrite` cannot change the logical size of a block. This is by design. See [`zfs-rewrite(8)`](https://openzfs.github.io/openzfs-docs/man/master/8/zfs-rewrite.8.html).

**Comment by Apachez-** [comment]
And there are no plans to let it be capable of changing the logical size of a block through syntax or through a new command like "zfs rebalance" or such?
